### PR TITLE
Stop generating metadata files at project's root

### LINKED ISSUE #618: Any way to get rid of .project .classpath .settings/ files?
Every time I open a Java project, no matter Maven or Gradle, the extension always creates many `.project` `.settings/` `.classpath` files in my project folder. And yes I know it's Eclipse's behavior to create files in project root folder as mentioned in https://github.com/redhat-developer/vscode-java/issues/466#issuecomment-373391208  But it's really very annoying that the extension creates these files silently without any confirmation, especially with the SCM stuffs, I have to manually add them into `.gitignore` file. And I also have to add them into "files.exclude" in VSCode's settings to hide them from the explorer.

I've noticed some projects are also created in the "workspaceStorage" folder, so **is it possible to create a corresponding project in the "workspaceStorage" folder and link my files to it?** I followed [this post](https://stackoverflow.com/questions/3422261/store-eclipse-project-files-outside-the-project-directory) in Eclipse, and find it adds `<linkedResources>` entries in `.project`. I suppose it might be somehow doable.

Moreover, **it would be also nice for me if it notifies the changes and provides shortcuts to hide/ignore them**, because I don't know and don't care about `.settings/` stuffs. E.g. it might pop up a message box saying something like "Some files are created by the extension, do you want to add them into .gitignore?"

### PR DESCRIPTION
**Pull request by jdneo**
Fixes #618, It requires https://github.com/eclipse/eclipse.jdt.ls/pull/1900

Signed-off-by: Sheng Chen <sheche@microsoft.com>

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -366,6 +366,12 @@
           ],
           "scope": "window"
         },
+        "java.import.generatesMetadataFilesAtProjectRoot": {
+          "type": "boolean",
+          "description": "Specify whether the project metadata files(.project, .classpath, .settings) will be generated at the project root.",
```
Comment (jdneo) [inline package.json:371]: Maybe we can add a wiki page link here to describe more background information about this setting

File: package.json [hunk 2]
```diff
@@ -366,6 +366,12 @@
           ],
           "scope": "window"
         },
+        "java.import.generatesMetadataFilesAtProjectRoot": {
```
Comment (testforstephen) [inline package.json:369]: Add new setting to README as well.

Comment (testforstephen) [reply package.json:369]: It requires to restart to take effect.

Need watch on the setting change event.
https://github.com/redhat-developer/vscode-java/blob/b4f8b07d774acb17bee19f3b06f4d0364a72c682/src/settings.ts#L35

Comment (testforstephen) [reply package.json:369]: It fails when changing the setting from `false` to `true`.

reproduce steps:
- keep this setting to default value `false`, then open `spring-petclinic`
- Once JLS is ready, then changing the setting to `true`.
- Restart VS Code.
- The project reimport fails.

`
!ENTRY org.eclipse.jdt.ls.core 4 0 2021-11-19 10:06:44.019!MESSAGE Failed to build workspace.!STACK 1Java Model Exception: Error in Java Model (code 969): jdt.ls-java-project does not existat org.eclipse.jdt.internal.core.JavaElement.newJavaModelException(JavaElement.java:584)at org.eclipse.jdt.internal.core.Openable.generateInfos(Openable.java:254)at org.eclipse.jdt.internal.core.JavaElement.openWhenClosed(JavaElement.java:597)at org.eclipse.jdt.internal.core.JavaElement.getElementInfo(JavaElement.java:328)at org.eclipse.jdt.internal.core.JavaElement.getElementInfo(JavaElement.java:314)at org.eclipse.jdt.internal.core.JavaElement.getChildren(JavaElement.java:269)at org.eclipse.jdt.internal.core.JavaProject.getPackageFragmentRoots(JavaProject.java:2355)at org.eclipse.jdt.internal.core.JavaProject.getPackageFragments(JavaProject.java:2379)at org.eclipse.jdt.ls.core.internal.managers.ProjectsManager.cleanupResources(ProjectsManager.java:234)at org.eclipse.jdt.ls.core.internal.handlers.BuildWorkspaceHandler.buildWorkspace(BuildWorkspaceHandler.java:53)at org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer.lambda$27(JDTLanguageServer.java:838)at org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer.lambda$53(JDTLanguageServer.java:1017)at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642)at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:479)at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)!SUBENTRY 1 org.eclipse.jdt.core 4 969 2021-11-19 10:06:44.019!MESSAGE jdt.ls-java-project does not exist
`

File: document/_java.metadataFilesGeneration.md [hunk 3]
```diff
@@ -0,0 +1,21 @@
+# Metadata Files Generation
+
+We use the setting `java.import.generatesMetadataFilesAtProjectRoot` to control where the project metadata files(.project, .classpath, .factorypath, .settings/) will be generated:
+- `true`: Metadata files will be generated at the project's root.
+- `false`: Metadata files will be generated at the workspace storage. To be specifc, the path will be: `<WORKSPACE_STORAGE_PATH>/redhat.java/jdt_ws/.metadata/.plugins/org.eclipse.core.resources/.projects/<PROJECT_NAME>`.
+
+By default, the setting is set to `false`.
+
+> If the metadata files exist both at the project root and the workspace storage, the extension will pick the project root ones.
```
Comment (testforstephen) [inline document/_java.metadataFilesGeneration.md:9]: If the metadata files exist in both the project root and the workspace storage, the extension will pick the files in the project root.

File: document/_java.metadataFilesGeneration.md [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+## Change the setting value
+
+Depending on how you change the setting, some extra steps need to be taken to make the change take effect.
+
+### Change from `false` to `true`
+You need to restart the client.
+
+### Chane from `true` to `false`
```
Comment (testforstephen) [inline document/_java.metadataFilesGeneration.md:18]: Change

**Review by testforstephen — APPROVED** [review]
LGTM

**Pull request merged**